stdlib/deplinker/elf.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3.6
+# -*- coding: utf-8 -*-
+"""A primordial dependency linker that assign requirements based on ELF dependencies."""
+
+import requests
+import ntpath
+import glob
+import braceexpand
+import urllib.parse
+import core.config
+import stdlib
+import stdlib.log
+from typing import Optional
+from elftools.elf.elffile import ELFFile
+
+
+def elf_deplinker(
+    packages,
+    search_patterns=[
+        '{,usr/}bin/*',
+        '{,usr/}lib{,32,64}/*',
+    ],
+    local_resolving: bool = True,
+    remote_resolving: bool = True,
+):
+    """A primordial dependency linker that assign requirements based on ELF dependencies.
+
+    This dependency linker will expand all ``search_patterns`` and try to analyse all files as ELF.
+
+    For all ELF, this dependency linker will retrieve all ELF dependencies. It will then see if any of those dependencies
+    are either provided by another package of the same build; or try to fetch the list of repositories
+    written in the :py:mod:`~core.config` file (in the order indicated) to find the unique package that contains
+    any of those files; and add it as a requirement.
+
+    :param packages: The packages which should have their dependencies assigned
+    :paramtype packages: ``List`` [ :py:class:`.Package` ]
+
+    :param search_patterns: A list of pattern which should match any ELF file present in the package's content.
+        Any element of this list supports the globbing and braces syntax of common shells.
+
+        The default value is: ::
+
+            [
+                '{,usr/}}bin/*',
+                '{,usr/}}lib{,32,64}/*',
+            ]
+
+        Which expands to: ::
+
+            [
+                'bin/*'
+                'usr/bin/*'
+                'lib/*'
+                'usr/lib/*'
+                'lib32/*'
+                'usr/lib32/*'
+                'lib64/*'
+                'usr/lib64/*'
+            ]
+
+    :paramtype search_patterns: ``List`` [ :py:class:`.Package` ]
+    :param local_resolving: Indicate whether or not other packages for ``packages`` should be used to solve dependencies. Default value is ``True``.
+    :param remote_resolving: Indicate whether or not remote repositories should be used to solve dependencies. Default value is ``True``.
+    """
+    binaries = dict()  # Key = a binary filename, Value = a package ID
+    dependencies = dict()  # Key = a package ID, Value = a list of ELF filename
+
+    for package in packages.values():
+        stdlib.log.ilog(f"ELF Dependency Linker: Looking for binaries and their dependencies in {package.id.full_name()}...")
+
+        pkg_id = package.id
+
+        # Find all elfs in this package
+        elfs = _find_elfs(package, search_patterns)
+
+        # Fill the `binaries` dict, to keep track of who owns a binary in the local context
+        #
+        # This generates unspecified behaviour if two packages have two binaries with the same filename
```

Comment:
You might want to `assert` in order to actually prevent unspecified behavior to happen silently